### ISSUE TITLE
Add '--diff' flag to show changes without modifying the source file

### ISSUE THREAD

**Issue by jamesquilty** [executablebooks/mdformat#268]
### Description / Summary

A nice-to-have feature would be for mdformat to show the changes it would make to the file by writing the output to stdout rather than to the original file. This is a feature that [black](https://github.com/psf/black) has, using the `--diff` flag on the command line.

### Value / benefit

At present, the only way to find out the changes that `mdformat` will make (or has made) is to create a copy of the file to be formatted, run `mdformat` on the copy and compare the two files. Or, if the file is under version control, to compare to the previous version (e.g. by `git diff`). The extra effort is a bit of a pest, but this feature is really just a nice-to-have.

**Comment by hukkin** [comment]
Thanks for the issue!

I'm not _too_ excited about this feature as
- it's pretty easy to print the diff by doing something like the following already
  ```sh
  diff foo.md <(mdformat - < foo.md)
  ```
- I find it more useful to know if a file is formatted or not, rather than what exact parts are not formatted. The best way to fix is by running `mdformat` anyways, I don't need concern myself with the formatting details.

I'll leave this open for now so we can reconsider if there's demand and real-life use cases from other users.

**Comment by jamesquilty** [comment]
Well, I'm not _too_ excited about the `--check` feature for exactly the same reasoning. YMMV. Black's implementation of this feature appears to be straightforward and I've added it to the **Implementation details** section above.

**Comment by psychemedia** [comment]
In passing, I note that `nbqa` [now supports](https://github.com/nbQA-dev/nbQA/issues/555#issuecomment-974651504) checks such as running `mdformat` over notebook Markdown cells, including the running of an `--nbqa-diff` switch that generates a preview rather than changing content in-place.

**Comment by veghdev** [comment]
I think a real use case could be that people submit a PR for documents in Markdown files via the Github web editor. If they make an error, CI will fail, but there is no way to tell the user what the formatting problem is. In many cases these people cannot use a development environment, but it is possible to fix their PRs with the diff output.

**Comment by MrMino** [comment]
@hukkin 

> it's pretty easy to print the diff by doing something like the following already

That assumes you are using 1) a shell 2) one that has the process substitution feature. Neither is the case if you are running it via `exec` call on a Docker container that uses Ash for its shell, i.e. in many CI flavors out there.

Lots of formatters have this feature exactly for this reason: so that there's an easy way to show the diff in CI logs.

**Comment by hukkin** [comment]
@MrMino good point. Let's add this. There is already PR https://github.com/executablebooks/mdformat/pull/381 which I'll find the time to review in the near future.
